feat: navigation component

kaljarv opened the pull request:
## WHY:

Necessary for #214 and related to #60.

### What has been changed (if possible, add screenshots, gifs, etc. )

The `Navigation` component takes as its input a list of `NavigationItems` and outputs a properly formatted navigation element with the items as links or buttons depending on whether a `href` or `onclick` property is provided for them.

The logic for creating the actual items is separated from the component so that the lists of `NavigationItems` can be contained in stores placed in the `$voter` and `$candidate` libraries.

The component will mainly be used in the page templates.

## Check off each of the following tasks as they are completed

- [x] I have reviewed the changes myself in this PR. Please check the [self-review document](https://github.com/OpenVAA/voting-advice-application/blob/main/docs/contributing/self-review.md)
- [ ] I have added or edited unit tests.
- [ ] I have run the unit tests successfully.
- [x] I have tested this change on my own device.
- [ ] I have tested this change on other devices (Using Browserstack is recommended).
- [x] I have tested my changes using the [WAVE extension](https://wave.webaim.org/extension/)
- [x] I have added documentation where necessary.
- [x] Is there an existing issue linked to this PR?

**Clean up your git commit history before submitting the pull request!**

kaljarv commented:
In order for the component to be reusable for both the Voter and the Candidate App, the logic of producing the actual navigation items is now separated from the layout.

Note that there are two `tmp:` commits that will be ditched before merging.

jaguarfi commented on frontend/src/lib/components/navigation/Navigation.svelte:
````diff
@@ -0,0 +1,72 @@
+```
+-->
+
+<nav {...$$restProps}>
````
I may have a big suggestion 😅: I think, we should have two components, the `Navigation` component, and a `NavItem` component. Something like this:

```Svelte
<script>
	import { Navbar, NavbarItem } from 'yesvelte'
</script>

<Navivation>
	<NavItem title="Home" />
	<NavItem title="Features" />
	<NavItem title="Pricing" />
	<NavItem title="Disabled" disabled />
</Navigation>
```

This way we can assign classes and the `aria` properties to each navigation Item. With the current implementation, it is not possible to pass properties to the navigation Items. 

Some inspiration:
https://mui.com/material-ui/react-menu/
https://www.yesvelte.com/docs/navbar

jaguarfi replied:
https://joyofcode.xyz/svelte-context-with-stores

kaljarv replied:
I'll convert it to multiple components! I think I'll also need a `NavGroup` which is basically a glorified `ul` :smile

jaguarfi reviewed (changes requested):
I just added a couple of comments. But I strongly believe we have to add a NavItem component. I could push the branch where I am refactoring the tabs so that you can see what I am trying to do. But, in the end, the new Tab component will look like this.
<img width="430" alt="image" src="https://github.com/OpenVAA/voting-advice-application/assets/10437475/31761156-3f9b-45ad-9e9b-349635824ecb">

With this approach, we can manage the aria properties easily, add styles to each subcomponent and make the component more dynamic.

Please let me know what you think about this.

kaljarv commented:
Okay, this version now has multiple subcomponents for building menus. I also added one for the Voter App.

@jaguarfi, is this something in the way you were thinking of?

jaguarfi commented on frontend/src/lib/components/navigation/NavItem.svelte:
```diff
@@ -0,0 +1,75 @@
+
+<!-- We use a div with an Aria role instead of a <li> because we don't
+  place the items within a valid parent for a <li> -->
+<div role="listitem">
```
shouldn't the role be `menuitem`

kaljarv replied:
See above.

jaguarfi commented on frontend/src/lib/voter/components/navigation/VoterNav.svelte:
```diff
@@ -0,0 +1,45 @@
+<script lang="ts">
```
We don't need this component. As the navigation for the voter app or the candidate app is defined only once, we can define it on the actual route page. Also, if we need to render some nav items dynamically based on the application state it will make more sense to define the Navigation menu in the pages.

kaljarv replied:
In my opinion, it depends a bit on how the page templates that will be added later are structured, but we can discuss this related to that PR once I get it started. Even there, though, I think it might make sense that navigation was its separate file, i.e., component, because it's logically a bit different from the page template itself and would make for nicer separation in terms of possible conflicts with changes to files.

Anyway, I can leave it out from this PR.

jaguarfi reviewed (changes requested):
A couple of comments regarding the roles for the components. We may need to also allow the navigation with the arrow keys.

https://www.w3.org/WAI/ARIA/apg/patterns/menubar/examples/menubar-navigation/

kaljarv commented on frontend/src/lib/components/navigation/NavItem.svelte:
```diff
@@ -0,0 +1,75 @@
+  type $$Props = NavItemProps;
+
+  export let icon: $$Props['icon'] = undefined;
+  export let href: $$Props['href'] = undefined;
+  export let onclick: $$Props['onclick'] = undefined;
```
Okay, we can do that. The only issue is that we cannot confirm whether button will actually do anything if we rely on `on:click` or a similar handler but I guess we just have to trust the developers in that 😄

kaljarv commented on frontend/src/lib/components/navigation/NavGroup.svelte:
```diff
@@ -0,0 +1,43 @@
+  invalid content for such an element. See: 
+  https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/listitem_role -->
+
+<section role="list" {...$$restProps}>
```
The ARIA guidelines say that a `menu` should contain more complex logic (think of the menu in an application) than just navigation.

kaljarv commented:
> We may need to also allow the navigation with the arrow keys.

I'm not sure if they will be very helpful here because the navigation menu is one-dimensional. Note that the linked example explicitly states that it is more complex than a typical navigation menu would be.

I'd suggest we make the arrow navigation optional and later implement that with a generic Svelte `use` directive, which can be applied in other similar situations as well.

kaljarv commented:
Update: I separated the `a`/ `button` logic from `NavItem` and moved it into its own component, bc that will be used also for button components, which are unrelated to navigation.

Also, I noticed that the `HTMLAttributes<T>` prop definitions do not actually contain all the possible properties and converted those to the correct `SvelteHTMLElements['nav']` format. This change will need to be made to all other typings using that format as well.

johannalaitila commented on frontend/src/lib/components/navigation/NavItem.svelte:
```diff
@@ -0,0 +1,59 @@
+<script lang="ts">
+  import {error} from '@sveltejs/kit';
```
Is this required anymore? See the Github warning

kaljarv replied:
True! I saw the warning as well and forgot to update this :D It's now corrected. I also improved the documentation a tiny bit.

kaljarv commented:
Based on discussion with @jaguarfi, I'll remove the `ActionItem` component and merge the logic into `NavItem`. The same kind of logic will be built into the buttons PR in a base component for them so that these two PRs don't depend on each other.

jaguarfi reviewed (approved):
Please remove the commits that are not part of this PR.

Sidenote: The keyboard navigation may need to be improved but it should be fine for now.
